### PR TITLE
roachtest: test selection is broken due to CategoriseTests returning map

### LINKED ISSUE #130329: roachtest: issue wth test selection as categorise test is returning a map
Test selector returns the list of all the tests from snowflake in a sequence sorted by the last run date. This must be maintained to ensure that we have a correct selection of successful test runs.
In the recent change, the CategoriseTests is returning a map instead of a list, which caused this selection process to break.

Jira issue: CRDB-42002

### PR DESCRIPTION
**Pull request by nameisbhaskar**
In a recent change on ensuring that test selection dos not cause starvation due to decommissioned tests, a regression was introduced. In this change, CategoriseTests function returned a map instead of a list. Because of this, the sequence in which the snowflake was returning data was no longer preserved. The snowflake query sorts the data by last run. This is important for the test selection of successful tests with 35% selection criteria.

This fix is to change the return of CategoriseTests to be a list.

Fixes: #130329
Epic: None

### CODE REVIEW

File: pkg/cmd/roachtest/main.go [hunk 1]
```diff
@@ -320,12 +320,23 @@ func updateSpecForSelectiveTests(ctx context.Context, specs []registry.TestSpec)
 	// Now, we want to take the tests common to both. These are the tests from which we need to select
 	// "successfulTestsSelectPct" percent tests to run.
 	successfulTests := make([]*testselector.TestDetails, 0)
-	for _, spec := range specs {
-		if td, ok := allTests[spec.Name]; ok && !td.Selected {
+
+	// specsMap contains a map of all the test names that are in specs
+	specsMap := make(map[string]struct{})
```
[the remark this answers is not in the document]
Comment (DarrylWong) [reply pkg/cmd/roachtest/main.go:325]: Although I could see how this is hard to reason about correctness. Something about needing so many data structures still seems off to me though.

File: pkg/cmd/roachtest/main.go [hunk 2]
```diff
@@ -320,12 +320,23 @@ func updateSpecForSelectiveTests(ctx context.Context, specs []registry.TestSpec)
...
+	// allTestsMap is maintained to check for the test details while skipping a test
+	allTestsMap := make(map[string]*testselector.TestDetails)
+	for i := 0; i < len(allTests); i++ {
+		td := allTests[i]
+		if _, ok := specsMap[td.Name]; ok && !td.Selected {
 			// adding only the unselected tests that are part of the specs
 			// These are tests that have been running successfully
 			successfulTests = append(successfulTests, td)
 		}
+		allTestsMap[td.Name] = td
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/main.go:339]: It's okay to still set this even if the test isn't in specs right? Because it should just be ignored anyway?

**Review by DarrylWong — COMMENTED** [review]
Can we be a bit more specific about why the ordering needs to be preserved i.e.  mention snowflake sorts it by last run? I know it's in the documentation, but I think it'd be good to mention it directly in the code/issue for clarity.

File: pkg/cmd/roachtest/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/main.go:325]: I can set it to nil, to be safe of any condition added:
for _, test := range specs {
		allTestsMap[test.Name] = nil
	}

**Comment by nameisbhaskar** [comment]
Thanks @DarrylWong for reviewing the change. I have made the changes. PTAL.

File: pkg/cmd/roachtest/main.go [hunk 3]
```diff
@@ -334,13 +345,13 @@ func updateSpecForSelectiveTests(ctx context.Context, specs []registry.TestSpec)
...
 	for i := range specs {
-		if testShouldBeSkipped(allTests, specs[i], roachtestflags.Suite) {
+		if testShouldBeSkipped(allTestsMap, specs[i], roachtestflags.Suite) {
 			specs[i].Skip = "test selector"
 			specs[i].SkipDetails = "test skipped because it is stable and selective-tests is set."
 		} else {
 			selectedTestsCount++
 		}
-		if td, ok := allTests[specs[i].Name]; ok {
+		if td, ok := allTestsMap[specs[i].Name]; ok && td != nil {
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/main.go:354]: What's the scenario that td can be nil? If we added a new test to the suite that doesn't have data in snowflake yet?

File: pkg/cmd/roachtest/main.go [hunk 4]
```diff
@@ -361,8 +372,8 @@ func testShouldBeSkipped(
 		return false
 	}
 
-	td, ok := testNamesToRun[test.Name]
-	return ok && test.Skip == "" && !td.Selected
+	td := testNamesToRun[test.Name]
+	return test.Skip == "" && td != nil && !td.Selected
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/main.go:376]: nit: I understand why we check for the `test.Skip == ""` condition here, so we don't overwrite existing skip details. But I think it's confusing to return that a test shouldn't be skipped if the test is marked with skip. Seems like it should be the opposite to me? Maybe we can extract the `test.Skip == ""` logic one layer to when we actually do the overwriting of `test.skip`?

File: pkg/cmd/roachtest/main.go (hunk 3 above)
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/main.go:354]: yes

File: pkg/cmd/roachtest/main.go (hunk 4 above)
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/main.go:376]: Sure! I can do this!

**Review by DarrylWong — COMMENTED** [review]
Okay, I think I follow the selection logic now. In general though, I feel like every time I revisit this code it takes me a while to re-reason through what each part is doing because there's so many edge cases to account for.

What do you think about adding a comment at the top that walks through at a high level what this function does step by step? Very rough outline, but something like this:

1. Query snowflake for previous test run data.
2. Mark tests that match certain criteria as selected. Remaining tests are sorted in order of last run date.
3. Find a list of `successfulTests` ordered by last run date. These are tests that have not been already marked as selected but are eligible (in `specs`) to be selected as part of the `SuccessfulTestsSelectPct` run.
4. Select `SuccessfulTestsSelectPct` of `successfulTests` to be run.
5. Select tests that are opted out of test selection to be run.
6. Mark all tests that are not selected and aren't new tests as skipped in the test spec.

Just a rough 2 minute draft but I think something like that would've helped me reason through what's happening a lot easier. 

------
Example of how writing that out helped me reason through the logic:

Shouldn't step 5 happen before step 3? Otherwise tests marked as opted out could still be selected as part of successful tests, even though they will be run anyway. 

I guess after one run, the opted out tests will always be at the bottom of the snowflake query and we won't run into it anymore, so maybe it's okay? Let me know what you think, besides that this LGTM.

**Review by DarrylWong — APPROVED** [review]
Thanks for adding all the comments!

If it's not too hard, would be nice to fix up the WIP unit test and stress this change a bit before merging.

**Comment by nameisbhaskar** [comment]
I have tested this change to check it returns the right tests. Thanks for reviewing this @DarrylWong !

**Pull request merged**